File: akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/BlogPostExample.java
/*
 * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.akka.persistence.typed;

import akka.Done;
import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.Behaviors;
import akka.persistence.typed.PersistenceId;
import akka.persistence.typed.javadsl.*;

public interface BlogPostExample {

  // #event
  interface BlogEvent {}

  public class PostAdded implements BlogEvent {
    private final String postId;
    private final PostContent content;

    public PostAdded(String postId, PostContent content) {
      this.postId = postId;
      this.content = content;
    }
  }

  public class BodyChanged implements BlogEvent {
    private final String postId;
    private final String newBody;

    public BodyChanged(String postId, String newBody) {
      this.postId = postId;
      this.newBody = newBody;
    }
  }

  public class Published implements BlogEvent {
    private final String postId;

    public Published(String postId) {
      this.postId = postId;
    }
  }
  // #event

  // #state
  interface BlogState {}

  public enum BlankState implements BlogState {
    INSTANCE
  }

  public class DraftState implements BlogState {
    final PostContent content;

    DraftState(PostContent content) {
      this.content = content;
    }

    public DraftState withContent(PostContent newContent) {
      return new DraftState(newContent);
    }

    public DraftState withBody(String newBody) {
      return withContent(new PostContent(postId(), content.title, newBody));
    }

    public String postId() {
      return content.postId;
    }
  }

  public class PublishedState implements BlogState {
    final PostContent content;

    PublishedState(PostContent content) {
      this.content = content;
    }

    public PublishedState withContent(PostContent newContent) {
      return new PublishedState(newContent);
    }

    public PublishedState withBody(String newBody) {
      return withContent(new PostContent(postId(), content.title, newBody));
    }

    public String postId() {
      return content.postId;
    }
  }
  // #state

  // #commands
  public interface BlogCommand {}
  // #reply-command
  public class AddPost implements BlogCommand {
    final PostContent content;
    final ActorRef<AddPostDone> replyTo;

    public AddPost(PostContent content, ActorRef<AddPostDone> replyTo) {
      this.content = content;
      this.replyTo = replyTo;
    }
  }

  public class AddPostDone implements BlogCommand {
    final String postId;

    public AddPostDone(String postId) {
      this.postId = postId;
    }
  }
  // #reply-command
  public class GetPost implements BlogCommand {
    final ActorRef<PostContent> replyTo;

    public GetPost(ActorRef<PostContent> replyTo) {
      this.replyTo = replyTo;
    }
  }

  public class ChangeBody implements BlogCommand {
    final String newBody;
    final ActorRef<Done> replyTo;

    public ChangeBody(String newBody, ActorRef<Done> replyTo) {
      this.newBody = newBody;
      this.replyTo = replyTo;
    }
  }

  public class Publish implements BlogCommand {
    final ActorRef<Done> replyTo;

    public Publish(ActorRef<Done> replyTo) {
      this.replyTo = replyTo;
    }
  }

  public class PostContent implements BlogCommand {
    final String postId;
    final String title;
    final String body;

    public PostContent(String postId, String title, String body) {
      this.postId = postId;
      this.title = title;
      this.body = body;
    }
  }
  // #commands

  // #behavior
  public class BlogBehavior extends EventSourcedBehavior<BlogCommand, BlogEvent, BlogState> {

    public BlogBehavior(PersistenceId persistenceId) {
      super(persistenceId);
    }

    // #behavior

    // #command-handler
    @Override
    public CommandHandler<BlogCommand, BlogEvent, BlogState> commandHandler() {
      CommandHandlerBuilder<BlogCommand, BlogEvent, BlogState> builder = newCommandHandlerBuilder();

      builder.forStateType(BlankState.class).onCommand(AddPost.class, this::addPost);

      builder
          .forStateType(DraftState.class)
          .onCommand(ChangeBody.class, this::changeBody)
          .onCommand(Publish.class, this::publish)
          .onCommand(GetPost.class, this::getPost);

      builder
          .forStateType(PublishedState.class)
          .onCommand(ChangeBody.class, this::changeBody)
          .onCommand(GetPost.class, this::getPost);

      builder.forAnyState().onCommand(AddPost.class, (state, cmd) -> Effect().unhandled());

      return builder.build();
    }

    private Effect<BlogEvent, BlogState> addPost(AddPost cmd) {
      // #reply
      PostAdded event = new PostAdded(cmd.content.postId, cmd.content);
      return Effect()
          .persist(event)
          .thenRun(() -> cmd.replyTo.tell(new AddPostDone(cmd.content.postId)));
      // #reply
    }

    private Effect<BlogEvent, BlogState> changeBody(DraftState state, ChangeBody cmd) {
      BodyChanged event = new BodyChanged(state.postId(), cmd.newBody);
      return Effect().persist(event).thenRun(() -> cmd.replyTo.tell(Done.getInstance()));
    }

    private Effect<BlogEvent, BlogState> changeBody(PublishedState state, ChangeBody cmd) {
      BodyChanged event = new BodyChanged(state.postId(), cmd.newBody);
      return Effect().persist(event).thenRun(() -> cmd.replyTo.tell(Done.getInstance()));
    }

    private Effect<BlogEvent, BlogState> publish(DraftState state, Publish cmd) {
      return Effect()
          .persist(new Published(state.postId()))
          .thenRun(
              () -> {
                System.out.println("Blog post published: " + state.postId());
                cmd.replyTo.tell(Done.getInstance());
              });
    }

    private Effect<BlogEvent, BlogState> getPost(DraftState state, GetPost cmd) {
      cmd.replyTo.tell(state.content);
      return Effect().none();
    }

    private Effect<BlogEvent, BlogState> getPost(PublishedState state, GetPost cmd) {
      cmd.replyTo.tell(state.content);
      return Effect().none();
    }
    // #command-handler

    // #event-handler
    @Override
    public EventHandler<BlogState, BlogEvent> eventHandler() {

      EventHandlerBuilder<BlogState, BlogEvent> builder = newEventHandlerBuilder();

      builder
          .forStateType(BlankState.class)
          .onEvent(PostAdded.class, event -> new DraftState(event.content));

      builder
          .forStateType(DraftState.class)
          .onEvent(BodyChanged.class, (state, chg) -> state.withBody(chg.newBody))
          .onEvent(Published.class, (state, event) -> new PublishedState(state.content));

      builder
          .forStateType(PublishedState.class)
          .onEvent(BodyChanged.class, (state, chg) -> state.withBody(chg.newBody));

      return builder.build();
    }
    // #event-handler

    // #behavior
    public static Behavior<BlogCommand> behavior(String entityId) {
      return Behaviors.setup(ctx -> new BlogBehavior(new PersistenceId("Blog-" + entityId)));
    }

    @Override
    public BlogState emptyState() {
      return BlankState.INSTANCE;
    }

    // commandHandler, eventHandler as in above snippets
  }
  // #behavior
}

File: akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/BlogPostExample.scala
/*
 * Copyright (C) 2017-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.akka.persistence.typed

import akka.Done
import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.persistence.typed.PersistenceId
import akka.persistence.typed.scaladsl.Effect
import akka.persistence.typed.scaladsl.EventSourcedBehavior

object BlogPostExample {

  //#event
  sealed trait BlogEvent
  final case class PostAdded(postId: String, content: PostContent) extends BlogEvent

  final case class BodyChanged(postId: String, newBody: String) extends BlogEvent
  final case class Published(postId: String) extends BlogEvent
  //#event

  //#state
  sealed trait BlogState

  case object BlankState extends BlogState

  final case class DraftState(content: PostContent) extends BlogState {
    def withBody(newBody: String): DraftState =
      copy(content = content.copy(body = newBody))

    def postId: String = content.postId
  }

  final case class PublishedState(content: PostContent) extends BlogState {
    def postId: String = content.postId
  }
  //#state

  //#commands
  sealed trait BlogCommand
  //#reply-command
  final case class AddPost(content: PostContent, replyTo: ActorRef[AddPostDone]) extends BlogCommand
  final case class AddPostDone(postId: String)
  //#reply-command
  final case class GetPost(replyTo: ActorRef[PostContent]) extends BlogCommand
  final case class ChangeBody(newBody: String, replyTo: ActorRef[Done]) extends BlogCommand
  final case class Publish(replyTo: ActorRef[Done]) extends BlogCommand
  final case class PostContent(postId: String, title: String, body: String)
  //#commands

  //#behavior
  def behavior(entityId: String): Behavior[BlogCommand] =
    EventSourcedBehavior[BlogCommand, BlogEvent, BlogState](
      persistenceId = PersistenceId(s"Blog-$entityId"),
      emptyState = BlankState,
      commandHandler,
      eventHandler)
  //#behavior

  //#command-handler
  private val commandHandler: (BlogState, BlogCommand) => Effect[BlogEvent, BlogState] = { (state, command) =>
    state match {

      case BlankState =>
        command match {
          case cmd: AddPost => addPost(cmd)
          case _            => Effect.unhandled
        }

      case draftState: DraftState =>
        command match {
          case cmd: ChangeBody  => changeBody(draftState, cmd)
          case Publish(replyTo) => publish(draftState, replyTo)
          case GetPost(replyTo) => getPost(draftState, replyTo)
          case _: AddPost       => Effect.unhandled
        }

      case publishedState: PublishedState =>
        command match {
          case GetPost(replyTo) => getPost(publishedState, replyTo)
          case _                => Effect.unhandled
        }
    }
  }

  private def addPost(cmd: AddPost): Effect[BlogEvent, BlogState] = {
    //#reply
    val evt = PostAdded(cmd.content.postId, cmd.content)
    Effect.persist(evt).thenRun { _ =>
      // After persist is done additional side effects can be performed
      cmd.replyTo ! AddPostDone(cmd.content.postId)
    }
    //#reply
  }

  private def changeBody(state: DraftState, cmd: ChangeBody): Effect[BlogEvent, BlogState] = {
    val evt = BodyChanged(state.postId, cmd.newBody)
    Effect.persist(evt).thenRun { _ =>
      cmd.replyTo ! Done
    }
  }

  private def publish(state: DraftState, replyTo: ActorRef[Done]): Effect[BlogEvent, BlogState] = {
    Effect.persist(Published(state.postId)).thenRun { _ =>
      println(s"Blog post ${state.postId} was published")
      replyTo ! Done
    }
  }

  private def getPost(state: DraftState, replyTo: ActorRef[PostContent]): Effect[BlogEvent, BlogState] = {
    replyTo ! state.content
    Effect.none
  }

  private def getPost(state: PublishedState, replyTo: ActorRef[PostContent]): Effect[BlogEvent, BlogState] = {
    replyTo ! state.content
    Effect.none
  }
  //#command-handler

  //#event-handler
  private val eventHandler: (BlogState, BlogEvent) => BlogState = { (state, event) =>
    state match {

      case BlankState =>
        event match {
          case PostAdded(_, content) =>
            DraftState(content)
          case _ => throw new IllegalStateException(s"unexpected event [$event] in state [$state]")
        }

      case draftState: DraftState =>
        event match {

          case BodyChanged(_, newBody) =>
            draftState.withBody(newBody)

          case Published(_) =>
            PublishedState(draftState.content)

          case _ => throw new IllegalStateException(s"unexpected event [$event] in state [$state]")
        }

      case _: PublishedState =>
        // no more changes after published
        throw new IllegalStateException(s"unexpected event [$event] in state [$state]")
    }
  }
  //#event-handler

}
